controlplane/kubeadm/controllers/helpers.go

```diff
@@ -141,18 +142,24 @@ func (r *KubeadmControlPlaneReconciler) cloneConfigsAndGenerateMachine(ctx conte
 	})
 	if err != nil {
 		// Safe to return early here since no resources have been created yet.
+		conditions.MarkFalse(kcp, controlplanev1.ControlPlaneInitializedCondition, controlplanev1.InfrastructureResourceCreationFailedReason,
```

Comment:
This method (`cloneConfigsAndGenerateMachine`) is called for both init and join, so we can't use a single `ControlPlaneInitializedCondition` here. I think we're going to need to think about this some more and how we report cloning failures. cc @fabriziopandini 

Also, see #3798 for a discussion on ControlPlaneInitialized

Reply:
Can we return an error wrapped with Condition data, and give the responsibility up the stack to mark the conditions? Something like [this](https://gist.github.com/shysank/80d83ce12a48a7db6051bcc44c43d2cc).

Reply:
I would prefer not to do this if we can avoid it. We recently worked to remove our RequeueAfterError, which is somewhat similar to this concept.

Reply:
Just want to get a discussion going. Apologies for not getting on this earlier. So, based on @ncdc's comments, and #3798 , it looks like `ControlPlaneInitializedCondition` is not the right condition to be used here because:
1. `cloneConfigsAndGenerateMachine ` is used for both kcp init, and scale up
2. `ControlPlaneInitializedCondition` has to do more with kubeadm creating `kubeadm-config` config map for cluster configuration.

I think there are two ways we can report cloning failures:

1. Use existing conditions with the message being the error returned from  `cloneConfigsAndGenerateMachine`. Eg. `Available` condition in case of kcp init. `ResizedCondition` in case of scale up (which is also used in cluster upgrade)
  **+**: 
    -  No change in API

   **-**:
    - Maybe not a very good ux experience, since the user has to parse the error message to find out if it's a cloning 
       error
     - Old condition may get overwritten. Eg. Scale up condition message that informs about the difference in number 
            in replicas  
  

2. Add a new condition, something like `MachineSpecNotGeneratedCondition` to report errors while cloning the templates. 
  **+**: 
    - Better ux, easy to parse, old conditions are left alone

   **-**:
    - API change
  
Sorry if I'm being a bit verbose, they are mostly for my understanding :) Please let me know wdyt @ncdc  @vincepri @fabriziopandini

Reply:
I prefer option 2. Not sure about that name though. Would we want 1 condition for the infra cloning, 1 for bootstrap cloning, and 1 for machine creation? Or 1 that applies for all 3?

Reply:
I thought we could use 1 condition, and use different reasons for each one. I can't think of a good name though. Maybe `TemplatesCloned`? Or just keep it simple, and have a 1 to 1 mapping with 3 conditions as `InfraTemplateCloned`, ``BootstrapTemplateCloned`, and `MachineCreated`?

Reply:
Maybe we could do just 1, and call it `MachinesCreated`. If it's false and an error severity, the reasons/messages can include the relevant information (such as "error cloning bootstrap template: <details>", "error clone infrastructure template: <details>", "error creating machine"). The reason I'm thinking it should be `MachinesCreated` (plural) is that we don't expose individual conditions for each replica, but we frequently do have multiple replicas, so this hopefully covers all the cases.

Reply:
Yeah that makes sense. Also, to confirm, we don't want to include errors that may occur during cleanup because it doesn't add any value in ux, right?

Reply:
So basically one of these?
- clone infra, fail to clone bootstrap, cleanup (which fails)
- clone infra, clone bootstrap, fail to create machine, cleanup (which fails)

Maybe we can hold off putting that in conditions?

Reply:
Yeah, it doesn't give any useful information for the user to take action on. I think we can hold it off for later.